Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+        }
+        
+        if flags.contains(.endHeaders) {
+            // no continuations following, so emit a full frame
+            frame.flags.formUnion(.endHeaders)
+            self.currentHeaderFrame = nil
+            return frame
+        } else {
+            self.currentHeaderFrame = frame
+            return nil
+        }
+    }
+}
+
+class HTTP2FrameEncoder {
+    private let allocator: ByteBufferAllocator
+    
+    init(allocator: ByteBufferAllocator) {
+        self.allocator = allocator
+    }
+    
+    /// Encodes the frame and optionally returns one or more blobs of data
+    /// ready for the system.
+    ///
+    /// Returned data blobs would include anything of potentially flexible
+    /// length, such as DATA payloads, header fragments in HEADERS or PUSH_PROMISE
+    /// frames, and so on. This is to avoid manually copying chunks of data which
+    /// we could just enqueue separately in sequence on the channel. Generally, if
+    /// we have a byte buffer somewhere, we will return that separately rather than
+    /// copy it into another buffer, with the corresponding allocation overhead.
+    ///
+    /// - Parameters:
+    ///   - frame: The frame to encode.
+    ///   - buf: Destination buffer for the encoded frame.
+    /// - Returns: An array containing zero or more additional buffers to send, in
+    ///            order. These may contain data frames' payload bytes, encoded
+    ///            header fragments, etc.
+    /// - Throws: Errors returned from HPACK encoder.
+    func encode(frame: HTTP2Frame, to buf: inout ByteBuffer) throws -> [IOData] {
+        // note our starting point
+        let start = buf.writerIndex
+        
+//      +-----------------------------------------------+
+//      |                 Length (24)                   |
+//      +---------------+---------------+---------------+
+//      |   Type (8)    |   Flags (8)   |
+//      +-+-------------+---------------+-------------------------------+
+//      |R|                 Stream Identifier (31)                      |
+//      +=+=============================================================+
+//      |                   Frame Payload (0...)                      ...
+//      +---------------------------------------------------------------+
+        
+        // skip 24-bit length for now, we'll fill that in later
+        buf.moveWriterIndex(forwardBy: 3)
+        
+        // 8-bit type
+        buf.write(integer: frame.payload.code)
+        // 8-bit flags -- we don't use padding when encoding in NIO, though
+        buf.write(integer: frame.flags.subtracting(.padded).rawValue)
+        // 32-bit stream identifier -- ensuring the top bit is empty
+        buf.write(integer: frame.streamID.safeNetworkStreamID ?? 0)
+        
+        func writePayloadSize(_ size: Int) {
+            var bytes: (UInt8, UInt8, UInt8)
+            bytes.0 = UInt8((size & 0xff_00_00) >> 16)
```

Comment:
Any reason not to just issue `buf.set(integer:)` calls for each of these, instead of having the temporary tuple?

Reply:
It looked like `set(integer:)` was doing a bunch of things based on the size of the integer, which seemed a bit much for a single byte. Writing these as a `ContiguousCollection` looked much faster, and as part of every frame header it seemed a logical place to think about that sort of optimization. Of course, I'm not entirely clear how the `set(integer:)` implementation will compile down and optimize, but it ultimately goes through some endianness logic and a memory copy even for a single byte, so…

Reply:
Hrm. So the endianness and the memory copy is a bit of a red herring (the endianness swap operation for UInt8 is inlinable and a no-op, so that all just goes away). However, in my quick experiment I noticed that actually doing this as three set ops leads to hilariously more complex assembly than doing it as a single operation on a tuple. There is no good reason for this: I consider that a failure of the optimiser in Swift, which should be able to spot that these are two utterly equivalent operations. I'll investigate further, but for now we can leave this approach in place.